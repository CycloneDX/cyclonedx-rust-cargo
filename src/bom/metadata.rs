use std::{io, marker::PhantomData};

use cargo::core::Package;
use chrono::{DateTime, Utc};
use serde::Serialize;
use xml_writer::XmlWriter;

use crate::ToXml;

#[derive(Serialize)]
pub struct Metadata<'a> {
    timestamp: DateTime<Utc>,
    #[serde(skip)]
    temp: PhantomData<&'a ()>,
}

impl<'a> Default for Metadata<'a> {
    fn default() -> Self {
        Self {
            timestamp: Utc::now(),
            temp: PhantomData::default(),
        }
    }
}

impl<'a> From<&'a Package> for Metadata<'a> {
    fn from(_pkg: &'a Package) -> Self {
        Default::default()
    }
}

impl ToXml for Metadata<'_> {
    fn to_xml<W: io::Write>(&self, xml: &mut XmlWriter<W>) -> io::Result<()> {
        xml.begin_elem("metadata")?;

        xml.elem_text(
            "timestamp",
            &self
                .timestamp
                .to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
        )?;

        xml.end_elem()
    }
}
