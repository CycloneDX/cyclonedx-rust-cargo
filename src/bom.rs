/*
 * This file is part of CycloneDX Rust Cargo.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * Copyright (c) OWASP Foundation. All Rights Reserved.
 */
use std::io;
use std::iter::FromIterator;

use cargo::core::Package;
use serde::{Serialize, Serializer};
use uuid::Uuid;
use xml_writer::XmlWriter;

use crate::metadata::Metadata;
use crate::{Component, ToXml};

static SPEC_VERSION: &'static str = "1.3";

#[derive(Clone, Copy, Serialize)]
enum BomFormat {
    CycloneDX,
}

fn uuid_to_urn<S: Serializer>(uuid: &Uuid, serializer: S) -> Result<S::Ok, S::Error> {
    serializer.collect_str(&uuid.to_urn())
}

/// A software bill of materials for a Rust crate.
#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Bom {
    bom_format: BomFormat,
    spec_version: &'static str,
    #[serde(serialize_with = "uuid_to_urn")]
    serial_number: Uuid,
    version: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Metadata>,
    components: Vec<Component>,
}

impl<'a> Default for Bom {
    fn default() -> Bom {
        Self {
            bom_format: BomFormat::CycloneDX,
            spec_version: SPEC_VERSION,
            version: 1,
            serial_number: Uuid::new_v4(),
            metadata: None::<Metadata>,
            components: Vec::new(),
        }
    }
}

/// Create a new BOM from a sequence of cargo package references.
impl<'a> FromIterator<&'a Package> for Bom {
    fn from_iter<T: IntoIterator<Item = &'a Package>>(iter: T) -> Self {
        Self {
            components: iter.into_iter().map(Component::from).collect(),
            metadata: Some(Metadata::default()),
            ..Default::default()
        }
    }
}

impl ToXml for Bom {
    fn to_xml<W: io::Write>(&self, xml: &mut XmlWriter<W>) -> io::Result<()> {
        let namespace = format!("http://cyclonedx.org/schema/bom/{}", SPEC_VERSION);
        xml.dtd("UTF-8")?;
        xml.begin_elem("bom")?;
        xml.attr("serialNumber", &self.serial_number.to_urn().to_string())?;
        xml.attr("version", "1")?;
        xml.attr("xmlns", namespace.as_str())?;

        if let Some(metadata) = &self.metadata {
            metadata.to_xml(xml)?;
        }

        xml.begin_elem("components")?;
        for component in &self.components {
            component.to_xml(xml)?;
        }
        xml.end_elem()?;

        xml.end_elem()
    }
}
