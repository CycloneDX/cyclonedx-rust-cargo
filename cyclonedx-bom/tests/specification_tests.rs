use cyclonedx_bom::models::bom::Bom;

#[test]
fn it_should_parse_all_of_the_valid_xml_specifications() {
    insta::glob!("data/1.3/valid*.xml", |path| {
        let file = std::fs::File::open(path).expect("Failed to read file: {path:?}");
        let bom = Bom::parse_from_xml_v1_3(file)
            .expect("Failed to parse the document as an SBOM: {path:?}");

        // TODO: validate here

        let mut output = Vec::new();
        bom.output_as_xml_v1_3(&mut output)
            .expect("Failed to output the file: {path:?}");
        let bom_output = String::from_utf8_lossy(&output).to_string();

        insta::assert_snapshot!(bom_output);
    });
}

#[test]
fn it_should_parse_all_of_the_valid_json_specifications() {
    insta::glob!("data/1.3/valid*.json", |path| {
        let file = std::fs::File::open(path).expect("Failed to read file: {path:?}");
        let bom = Bom::parse_from_json_v1_3(file)
            .expect("Failed to parse the document as an SBOM: {path:?}");

        // TODO: validate here

        let mut output = Vec::new();
        bom.output_as_json_v1_3(&mut output)
            .expect("Failed to output the file: {path:?}");
        let bom_output = String::from_utf8_lossy(&output).to_string();

        insta::assert_snapshot!(bom_output);
    });
}
