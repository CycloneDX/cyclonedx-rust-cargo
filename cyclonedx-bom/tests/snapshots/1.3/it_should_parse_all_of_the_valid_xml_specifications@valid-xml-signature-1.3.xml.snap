---
source: cyclonedx-bom/tests/specification_tests.rs
expression: bom_output
input_file: cyclonedx-bom/tests/data/1.3/valid-xml-signature-1.3.xml
---
<?xml version="1.0" encoding="utf-8"?>
<bom xmlns="http://cyclonedx.org/schema/bom/1.3" serialNumber="urn:uuid:3e671687-395b-41f5-a30f-a58921a69b79" version="1">
  <components>
    <component type="application">
      <publisher>Acme Inc</publisher>
      <group>com.acme</group>
      <name>tomcat-catalina</name>
      <version>9.0.14</version>
      <hashes>
        <hash alg="MD5">3942447fac867ae5cdb3229b658f4d48</hash>
        <hash alg="SHA-1">e6b1000b94e835ffd37f4c6dcbdad43f4b48a02a</hash>
        <hash alg="SHA-256">f498a8ff2dd007e29c2074f5e4b01a9a01775c3ff3aeaf6906ea503bc5791b7b</hash>
        <hash alg="SHA-512">e8f33e424f3f4ed6db76a482fde1a5298970e442c531729119e37991884bdffab4f9426b7ee11fccd074eeda0634d71697d6f88a460dce0ac8d627a29f7d1282</hash>
      </hashes>
      <licenses>
        <license>
          <id>Apache-2.0</id>
        </license>
      </licenses>
      <purl>pkg:maven/com.acme/tomcat-catalina@9.0.14?packaging=jar</purl>
      <pedigree>
        <ancestors>
          <component type="application">
            <publisher>Apache</publisher>
            <group>org.apache.tomcat</group>
            <name>tomcat-catalina</name>
            <version>9.0.14</version>
            <licenses>
              <license>
                <id>Apache-2.0</id>
              </license>
            </licenses>
            <purl>pkg:maven/org.apache.tomcat/tomcat-catalina@9.0.14?packaging=jar</purl>
          </component>
        </ancestors>
        <commits>
          <commit>
            <uid>7638417db6d59f3c431d3e1f261cc637155684cd</uid>
            <url>https://location/to/7638417db6d59f3c431d3e1f261cc637155684cd</url>
            <author>
              <timestamp>2018-11-07T22:01:45Z</timestamp>
              <name>John Doe</name>
              <email>jdoe@example.com</email>
            </author>
            <committer>
              <timestamp>2018-11-07T22:01:45Z</timestamp>
              <name>John Doe</name>
              <email>jdoe@example.com</email>
            </committer>
            <message>Initial commit</message>
          </commit>
        </commits>
      </pedigree>
    </component>
    <component type="library">
      <group>org.example</group>
      <name>mylibrary</name>
      <version>1.0.0</version>
      <scope>required</scope>
      <hashes>
        <hash alg="MD5">2342c2eaf1feb9a80195dbaddf2ebaa3</hash>
        <hash alg="SHA-1">68b78babe00a053f9e35ec6a2d9080f5b90122b0</hash>
        <hash alg="SHA-256">708f1f53b41f11f02d12a11b1a38d2905d47b099afc71a0f1124ef8582ec7313</hash>
        <hash alg="SHA-512">387b7ae16b9cae45f830671541539bf544202faae5aac544a93b7b0a04f5f846fa2f4e81ef3f1677e13aed7496408a441f5657ab6d54423e56bf6f38da124aef</hash>
      </hashes>
      <licenses>
        <license>
          <id>Apache-2.0</id>
          <text content-type="text/xml" encoding="base64">blah</text>
          <url>fdaf</url>
        </license>
      </licenses>
      <copyright>Copyright Example Inc. All rights reserved.</copyright>
      <cpe>cpe:/a:example:myapplication:1.0.0</cpe>
      <purl>pkg:maven/com.example/myapplication@1.0.0?packaging=war</purl>
      <modified>false</modified>
    </component>
    <component type="framework">
      <group>com.example</group>
      <name>myframework</name>
      <version>1.0.0</version>
      <description>Example Inc, enterprise framework</description>
      <scope>required</scope>
      <hashes>
        <hash alg="MD5">cfcb0b64aacd2f81c1cd546543de965a</hash>
        <hash alg="SHA-1">7fbeef2346c45d565c3341f037bce4e088af8a52</hash>
        <hash alg="SHA-256">0384db3cec55d86a6898c489fdb75a8e75fe66b26639634983d2f3c3558493d1</hash>
        <hash alg="SHA-512">854909cdb9e3ca183056837144aab6d8069b377bd66445087cc7157bf0c3f620418705dd0b83bdc2f73a508c2bdb316ca1809d75ee6972d02023a3e7dd655c79</hash>
      </hashes>
      <licenses>
        <license>
          <id>Apache-2.0</id>
        </license>
      </licenses>
      <purl>pkg:maven/com.example/myframework@1.0.0?packaging=war</purl>
      <modified>false</modified>
      <externalReferences>
        <reference type="website">
          <url>http://example.com/myframework</url>
        </reference>
        <reference type="advisories">
          <url>http://example.com/security</url>
        </reference>
      </externalReferences>
    </component>
  </components>
</bom>
